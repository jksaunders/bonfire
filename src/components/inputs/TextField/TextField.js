import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Box, { BoxPropTypes, BoxDefaultProps } from '../../containers/Box';
import Layout from '../../containers/Layout';
import Typography from '../../Typography';

const propTypes = {
  ...BoxPropTypes,
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      message: PropTypes.string,
      color: PropTypes.string,
    }),
  ]),
  name: PropTypes.string,
  onInput: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  showError: PropTypes.bool,
  validate: PropTypes.func,
};

const defaultProps = {
  ...BoxDefaultProps,
  name: null,
  onInput: null,
  onSubmit: null,
  placeholder: null,
  showError: true,
  validate: null,
};

const borderColor = 'rgba(102,119,136,.2)';

const StyledInput = styled(Box)`
  ::placeholder {
    color: ${borderColor};
    font-size: 1rem;
  }
`;

const getError = error => {
  if (typeof error === 'string') {
    return error;
  }
  return '';
};

const validateString = (string, validate, currentError, setError) => {
  if (validate == null) {
    return;
  }

  const error = validate(string);
  if (error !== currentError) {
    if (error != null) {
      setError(error);
    } else {
      setError(null);
    }
  }
};

const getOnInput = ({ onInput, validate, currentError, setError }) => {
  if (onInput == null && validate == null) {
    return null;
  }

  return e => {
    if (onInput != null) {
      onInput(e.target.value);
    }
    validateString(e.target.value, validate, currentError, setError);
  };
};

const getOnKeyUp = ({ onSubmit }) => {
  if (onSubmit == null) {
    return null;
  }

  return e => {
    if (e.keyCode === 13 || e.which === 13) {
      onSubmit();
    }
  };
};

const TextField = ({
  name,
  error: errorProp,
  onInput,
  onSubmit,
  placeholder,
  showError,
  validate,
  ...rest
}) => {
  const [error, setError] = useState(false);
  const inputRef = useRef(null);

  useEffect(() => {
    // check error on render as well as onInput
    if (
      inputRef != null &&
      inputRef.current != null &&
      validate != null &&
      errorProp == null &&
      error != null
    ) {
      validateString(inputRef.current.value, validate, error, setError);
    }
  }, [validate, errorProp, error]);

  return (
    <Layout gap="4px">
      <StyledInput
        {...rest}
        border={`1px solid ${borderColor}`}
        borderRadius="5px"
        component="input"
        padding="3px"
        type="text"
        name={name}
        onInput={getOnInput({ onInput, validate, error, setError })}
        onKeyUp={getOnKeyUp({ onSubmit })}
        placeholder={placeholder}
        ref={inputRef}
      />
      {(errorProp || error) && showError && (
        <Typography size="14px">{getError(errorProp || error)}</Typography>
      )}
    </Layout>
  );
};

TextField.propTypes = propTypes;
TextField.defaultProps = defaultProps;

export default TextField;
