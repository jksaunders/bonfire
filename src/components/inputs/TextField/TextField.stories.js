import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import Layout from '../../containers/Layout';
import TextField from './TextField';
import { TypographyContext, TrelloVariants } from '../../Typography';

export default {
  title: 'Inputs|TextField',
  component: TextField,
  decorators: [withKnobs],
};

export const TextFieldStory = () => {
  return (
    // eslint-disable-next-line no-console
    <TextField placeholder="Text field" onSubmit={() => console.log('Done!')} />
  );
};

TextFieldStory.story = {
  name: 'Text Field',
};

export const Styled = () => (
  <TypographyContext.Provider value={TrelloVariants.Body}>
    <Layout maxWidth="50%" useTypography>
      <TextField placeholder="Text field" />
    </Layout>
  </TypographyContext.Provider>
);
