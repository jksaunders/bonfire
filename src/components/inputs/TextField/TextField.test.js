import React from 'react';
import {
  expectExists,
  expectSnapshot,
  fireEvent,
  render,
} from '../../../testing';
import TextField from './TextField';

const placeholder = 'input placeholder';

const getTextField = (props = {}, children) =>
  render(
    <TextField {...props} placeholder={placeholder}>
      {children}
    </TextField>
  );

const typeLetters = (node, value) =>
  fireEvent.input(node, { target: { value } });

test('renders', () => {
  const { component } = getTextField();
  expectExists(component);
  expectSnapshot(component);
});

test('`data-test` prop', () => {
  const { component, getByPlaceholderText } = getTextField({
    'data-test': 'testId',
  });
  expectExists(component);
  expectSnapshot(component);
  expect(getByPlaceholderText(placeholder)).toHaveAttribute(
    'data-test',
    'testId'
  );
});

describe('as textarea', () => {
  test('renders', () => {
    const { component, getByPlaceholderText } = getTextField({
      textArea: true,
    });
    expectExists(component);
    expectSnapshot(component);
    expectExists(getByPlaceholderText(placeholder));
    expect(getByPlaceholderText(placeholder).tagName.toLowerCase()).toBe(
      'textarea'
    );
  });
});

describe('validation', () => {
  const validate = text => text.length === 5;

  describe('error is passed as a prop', () => {
    const errorMessage = 'test error';

    test('error is shown when input is valid', () => {
      const { component, getByText, getByPlaceholderText } = getTextField({
        validate,
        error: errorMessage,
      });

      getByText(errorMessage);
      expectSnapshot(component);

      const textField = getByPlaceholderText(placeholder);

      typeLetters(textField, '12345');

      expect(getByPlaceholderText(placeholder).value).toEqual('12345');
      expect(getByText(errorMessage)).toBeInTheDocument();
    });

    test.todo('error is shown when input is not valid');
  });
});
