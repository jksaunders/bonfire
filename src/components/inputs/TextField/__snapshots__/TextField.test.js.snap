// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as textarea renders 1`] = `
.c0 {
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
  border: 1px solid rgba(102,119,136,.2);
  border-radius: 5px;
  padding: 3px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
}

.c1 {
  display: grid;
  grid-gap: 4px;
  -webkit-font-smoothing: antialiased;
}

.c4 {
  width: 100%;
}

.c4::-webkit-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::-moz-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4:-ms-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

<div
  class="c0 c1"
>
  <label>
    <span
      class="c2"
    />
  </label>
  <textarea
    class="c3 c4"
    placeholder="input placeholder"
    type="text"
  />
</div>
`;

exports[`renders 1`] = `
.c0 {
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
  border: 1px solid rgba(102,119,136,.2);
  border-radius: 5px;
  padding: 3px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
}

.c1 {
  display: grid;
  grid-gap: 4px;
  -webkit-font-smoothing: antialiased;
}

.c4 {
  width: 100%;
}

.c4::-webkit-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::-moz-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4:-ms-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

<div
  class="c0 c1"
>
  <label>
    <span
      class="c2"
    />
  </label>
  <input
    class="c3 c4"
    placeholder="input placeholder"
    type="text"
  />
</div>
`;

exports[`validation error is passed as a prop error is shown when input is valid 1`] = `
.c0 {
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
  border: 1px solid rgba(102,119,136,.2);
  border-radius: 5px;
  padding: 3px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
}

.c1 {
  display: grid;
  grid-gap: 4px;
  -webkit-font-smoothing: antialiased;
}

.c4 {
  width: 100%;
}

.c4::-webkit-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::-moz-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4:-ms-input-placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

.c4::placeholder {
  color: rgba(102,119,136,.2);
  font-size: 1rem;
}

<div
  class="c0 c1"
>
  <label>
    <span
      class="c2"
    />
  </label>
  <input
    class="c3 c4"
    placeholder="input placeholder"
    type="text"
  />
  <span
    class="c2"
  >
    test error
  </span>
</div>
`;
